### PR TITLE
New repository structure.

### PR DESCRIPTION
**Pull request by francesco-romano**
- libraries: contains project-wise libraries (which must be built before any plugin)
- thrift: contains rpc targets. Each target should build to a separate library (with relative includes) so that it can be linked by clients
- plugins: contains the plugins

Fixes https://github.com/robotology/gazebo-yarp-plugins/issues/187

### CODE REVIEW

**Comment by francesco-romano** [comment]
cc @traversaro for a first check.

There is still some issues (which may be solved after merge, or not)
- [ ] `yarp_idl`: when it autogenerates the files it does not "respect" the namespace convention: a header file with namespace `gazebo` is not saved in filesystem as `include\gazebo\file.h`. This creates issues then in the inclusion.
- [ ] I started to use `target_include_directories` instead of exporting global variables. The only question here is: what is the priority for the headers? build or install dir?
- [ ] I created a separate library for the thrift (currently only the clock, but the logic remains the same): this should allow clients to properly link to the autogenerated library instead of copy-paste sources (see https://github.com/robotology-playground/WBI-Toolbox for the `ySynchronizer` block). But probably I'm missing the `config.cmake` file.

Any other suggestions are welcome :smile:

**Comment by francesco-romano** [comment]
Unfortunately I bumped the cmake required version (not explicitly yet) to 3.0.0

File: libraries/common/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+# CopyPolicy: Released under the terms of the LGPLv2.1 or later, see LGPL.TXT
+
+cmake_minimum_required(VERSION 2.8.11)
+
+PROJECT(Library_Common)
+
+add_library(gazebo_yarp_lib_common INTERFACE)
+target_include_directories(gazebo_yarp_lib_common INTERFACE
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+)
```
Comment (traversaro) [inline libraries/common/CMakeLists.txt]: Cool, I did not know you could use this trick for header only libraries.

**Comment by traversaro** [comment]
What is requiring CMake 3.0 ? This could be a problem on Ubuntu 12.04 and Debian 7.

**Comment by francesco-romano** [comment]
The trick I used for the header only library. I discovered this only after

**Comment by francesco-romano** [comment]
@traversaro I added the possibility to specify a plugin as a module (for now default is shared).
I removed the requirement on CMake 3.0

**Comment by traversaro** [comment]
I think that thee issues in https://github.com/robotology/gazebo-yarp-plugins/pull/188#issuecomment-98341893 can be solved also after the merge. 
The blocking problem that I see at this moment is that we are breaking the documentation. because is not possible anymore to add the build directory to the `GAZEBO_PLUGIN_PATH` environmental variable, because all the plugins are now scattered in several subdirectories. I guess this is ok, but we should update the documentation and send a warning to the codyco users that the installation procedure has changed.

**Comment by francesco-romano** [comment]
@traversaro I updated the documentation accordingly

**Comment by traversaro** [comment]
Perfect, then we can merge. :+1: 

Public announcement : 
@azadm @robertocalandra @barbalberto @EnricoMingo @MirkoFerrati @rlober @hu-yue @mikearagao 
Please notice that after this change, it is not really feasible to use the plugin directly from the build directory. Check the updated documentation for how to properly use the plugins from the install directory:
https://github.com/francesco-romano/gazebo-yarp-plugins/commit/51d052196b45a9c6efc9d168956ebdeaee6c3412
(Please propagate this information to your colleagues , thanks!)

**Pull request merged**